### ISSUE TITLE
[Enterprise]: Found two visible definitions when modifying argument in a changeset

### ISSUE THREAD

**Issue by mehmetdavut** [rmosolgo/graphql-ruby#4440]
**Describe the bug**
When we want to modify an argument in the actual schema by a changeset `modifies` [as described](https://graphql-ruby.org/changesets/definition.html#arguments) in the documentation.

It only happens with modifying an argument's type.

**Versions**

`graphql` version:
graphql-enterprise (1.2.0)
      graphql (>= 1.13.0)
      graphql-pro (>= 1.24.0)
    graphql-pro (1.24.0)
      graphql (>= 1.7.5)

`rails` (or other framework): (7.0.4.3)

**GraphQL schema**

Include relevant types and fields (in Ruby is best, in GraphQL IDL is ok). Any custom extensions, etc?

```ruby
# the argument in the actual schema
class Types::QueryType < Types::BaseObject
  field :a_field_in_the_schema, ATypeInTheSchema.connection_type, null: true do
    argument :argument_should_change, Integer, required: false
  end
end

# the argument in the changeset
modifies Types::QueryType do
  modifies :a_field_in_the_schema do
    argument :argument_should_change, Decimal, required: false
  end
end
```

**Steps to reproduce**
- Create a changeset
- The changeset should try to modify an argument's type

**Expected behavior**
- I would like to have the argument's type is modified when I enable the changeset

**Actual behavior**

Found two visible definitions for `Query.aFieldInTheSchema.argumentShouldChange`

```


lib/graphql/schema/warden.rb:58 block in visible_entry? |  
-- | --
lib/graphql/schema/warden.rb:53 each |  
lib/graphql/schema/warden.rb:53 visible_entry? |  
lib/graphql/schema/member/has_arguments.rb:116 block in arguments |  
lib/graphql/schema/member/has_arguments.rb:115 each |  
lib/graphql/schema/member/has_arguments.rb:115 arguments |  
lib/graphql/schema/member/has_arguments.rb:177 arguments |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/index_visitor.rb:63 block in on_argument |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/index_visitor.rb:106 block in each_field |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/index_visitor.rb:105 each |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/index_visitor.rb:105 each_field |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/index_visitor.rb:59 on_argument |  
lib/graphql/static_validation/rules/variable_usages_are_allowed.rb:49 on_argument |  
lib/graphql/static_validation/rules/argument_literals_are_compatible.rb:8 on_argument |  
lib/graphql/static_validation/rules/arguments_are_defined.rb:9 on_argument |  
lib/graphql/static_validation/base_visitor.rb:120 on_argument |  
lib/graphql/language/visitor.rb:118 on_argument_with_modifications |  
lib/graphql/language/visitor.rb:155 block in on_field_children |  
lib/graphql/language/visitor.rb:154 each |  
lib/graphql/language/visitor.rb:154 on_field_children |  
lib/graphql/language/visitor.rb:105 on_field |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/index_visitor.rb:49 on_field |  
lib/graphql/static_validation/rules/argument_names_are_unique.rb:9 on_field |  
lib/graphql/static_validation/rules/required_arguments_are_present.rb:7 on_field |  
lib/graphql/static_validation/rules/fields_have_appropriate_selections.rb:12 on_field |  
lib/graphql/static_validation/rules/fields_will_merge.rb:31 on_field |  
lib/graphql/static_validation/rules/fields_are_defined_on_type.rb:25 on_field |  
lib/graphql/static_validation/rules/unique_directives_per_location.rb:27 block (2 levels) in <module:UniqueDirectivesPerLocation> |  
lib/graphql/static_validation/base_visitor.rb:89 on_field |  
lib/graphql/language/visitor.rb:118 on_field_with_modifications |  
lib/graphql/language/visitor.rb:181 block in visit_selections |  
lib/graphql/language/visitor.rb:178 each |  
lib/graphql/language/visitor.rb:178 visit_selections |  
lib/graphql/language/visitor.rb:214 on_operation_definition_children |  
lib/graphql/language/visitor.rb:105 on_operation_definition |  
lib/graphql/static_validation/definition_dependencies.rb:47 on_operation_definition |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/operation_names_are_present_and_unique.rb:21 on_operation_definition |  
lib/graphql/static_validation/rules/subscription_root_exists.rb:12 on_operation_definition |  
lib/graphql/static_validation/rules/query_root_exists.rb:12 on_operation_definition |  
lib/graphql/static_validation/rules/mutation_root_exists.rb:12 on_operation_definition |  
lib/graphql/static_validation/rules/variable_usages_are_allowed.rb:13 on_operation_definition |  
lib/graphql/static_validation/rules/variables_are_used_and_defined.rb:44 on_operation_definition |  
lib/graphql/static_validation/rules/variable_names_are_unique.rb:20 on_operation_definition |  
lib/graphql/static_validation/rules/fields_have_appropriate_selections.rb:18 on_operation_definition |  
lib/graphql/static_validation/rules/fields_will_merge.rb:26 on_operation_definition |  
lib/graphql/static_validation/rules/unique_directives_per_location.rb:27 block (2 levels) in <module:UniqueDirectivesPerLocation> |  
lib/graphql/static_validation/base_visitor.rb:59 on_operation_definition |  
lib/graphql/language/visitor.rb:118 on_operation_definition_with_modifications |  
lib/graphql/language/visitor.rb:144 public_send |  
lib/graphql/language/visitor.rb:144 block in on_document_children |  
lib/graphql/language/visitor.rb:142 each |  
lib/graphql/language/visitor.rb:142 on_document_children |  
lib/graphql/language/visitor.rb:105 on_document |  
lib/graphql/static_validation/definition_dependencies.rb:38 on_document |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/operation_names_are_present_and_unique.rb:25 on_document |  
lib/graphql/static_validation/rules/variables_are_used_and_defined.rb:79 on_document |  
lib/graphql/static_validation/rules/fragment_spreads_are_possible.rb:26 on_document |  
lib/graphql/static_validation/rules/fragments_are_used.rb:6 on_document |  
lib/graphql/static_validation/rules/fragments_are_finite.rb:6 on_document |  
lib/graphql/static_validation/rules/fragment_names_are_unique.rb:17 on_document |  
lib/graphql/static_validation/rules/no_definitions_are_present.rb:34 on_document |  
lib/graphql/language/visitor.rb:118 on_document_with_modifications |  
lib/graphql/language/visitor.rb:70 public_send |  
lib/graphql/language/visitor.rb:70 visit |  
lib/graphql/static_validation/validator.rb:44 block (3 levels) in validate |  
lib/graphql/static_validation/validator.rb:43 catch |  
lib/graphql/static_validation/validator.rb:43 block (2 levels) in validate |  
lib/timeout.rb:173 timeout |  
lib/graphql/static_validation/validator.rb:42 block in validate |  
lib/graphql/tracing.rb:80 call_tracers |  
lib/graphql/tracing.rb:82 block in call_tracers |  
lib/graphql/tracing/platform_tracing.rb:30 block in trace |  
lib/graphql/tracing/appsignal_tracing.rb:34 block in platform_trace |  
lib/appsignal/helpers/instrumentation.rb:528 instrument |  
lib/graphql/tracing/appsignal_tracing.rb:33 platform_trace |  
lib/graphql/tracing/platform_tracing.rb:29 trace |  
lib/graphql/tracing.rb:82 call_tracers |  
lib/graphql/tracing.rb:82 block in call_tracers |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/query_instrumentation.rb:19 trace |  
lib/graphql/tracing.rb:82 call_tracers |  
lib/graphql/tracing.rb:66 trace |  
lib/graphql/static_validation/validator.rb:30 validate |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/validate.rb:41 validate |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/active_record_backend.rb:48 block in batch_upsert_client_operations |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/active_record_backend.rb:47 map |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/active_record_backend.rb:47 batch_upsert_client_operations |  
/usr/local/lib/ruby/3.1.0/forwardable.rb:238 batch_upsert_client_operations |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/add_operation_batch.rb:159 save! |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/add_operation_batch.rb:38 block in call |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/active_record_backend.rb:364 block in transaction |  
lib/active_record/connection_adapters/abstract/transaction.rb:319 block in within_new_transaction |  
lib/active_support/concurrency/load_interlock_aware_monitor.rb:25 handle_interrupt |  
lib/active_support/concurrency/load_interlock_aware_monitor.rb:25 block in synchronize |  
lib/active_support/concurrency/load_interlock_aware_monitor.rb:21 handle_interrupt |  
lib/active_support/concurrency/load_interlock_aware_monitor.rb:21 synchronize |  
lib/active_record/connection_adapters/abstract/transaction.rb:317 within_new_transaction |  
lib/active_record/connection_adapters/abstract/database_statements.rb:316 transaction |  
lib/active_record/transactions.rb:209 transaction |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/active_record_backend.rb:362 transaction |  
/usr/local/lib/ruby/3.1.0/forwardable.rb:238 transaction |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/add_operation_batch.rb:27 call |  
graphql-pro (1.24.0) lib/graphql/pro/operation_store/endpoint.rb:42 call |  
lib/action_dispatch/routing/mapper.rb:19 block in <class:Constraints> |  
lib/action_dispatch/routing/mapper.rb:48 serve |  
lib/action_dispatch/journey/router.rb:50 block in serve |  
lib/action_dispatch/journey/router.rb:32 each |  
lib/action_dispatch/journey/router.rb:32 serve |  
lib/action_dispatch/routing/route_set.rb:852 call |  
rack-attack (6.6.1) lib/rack/attack.rb:103 call |  
rack-attack (6.6.1) lib/rack/attack.rb:127 call |  
lib/warden/manager.rb:36 block in call |  
lib/warden/manager.rb:34 catch |  
lib/warden/manager.rb:34 call |  
lib/rack/tempfile_reaper.rb:15 call |  
lib/rack/etag.rb:27 call |  
lib/rack/conditional_get.rb:40 call |  
lib/rack/head.rb:12 call |  
lib/action_dispatch/http/permissions_policy.rb:38 call |  
lib/action_dispatch/http/content_security_policy.rb:36 call |  
lib/action_dispatch/middleware/callbacks.rb:27 block in call |  
lib/active_support/callbacks.rb:99 run_callbacks |  
lib/action_dispatch/middleware/callbacks.rb:26 call |  
lib/appsignal/rack/rails_instrumentation.rb:32 call_with_appsignal_monitoring |  
lib/appsignal/rack/rails_instrumentation.rb:17 call |  
lib/action_dispatch/middleware/debug_exceptions.rb:28 call |  
lib/action_dispatch/middleware/show_exceptions.rb:26 call |  
lib/lograge/rails_ext/rack/logger.rb:18 call_app |  
lib/rails/rack/logger.rb:25 block in call |  
lib/active_support/tagged_logging.rb:99 block in tagged |  
lib/active_support/tagged_logging.rb:37 tagged |  
lib/active_support/tagged_logging.rb:99 tagged |  
lib/rails/rack/logger.rb:25 call |  
lib/action_dispatch/middleware/remote_ip.rb:93 call |  
lib/request_store/middleware.rb:19 call |  
lib/action_dispatch/middleware/request_id.rb:26 call |  
lib/rack/method_override.rb:24 call |  
lib/rack/runtime.rb:22 call |  
lib/active_support/cache/strategy/local_cache_middleware.rb:29 call |  
lib/action_dispatch/middleware/executor.rb:14 call |  
lib/action_dispatch/middleware/static.rb:23 call |  
lib/rack/sendfile.rb:110 call |  
lib/action_dispatch/middleware/ssl.rb:77 call |  
lib/action_dispatch/middleware/host_authorization.rb:131 call |  
rack-cors (1.1.1) lib/rack/cors.rb:100 call |  
lib/rails/engine.rb:530 call |  
lib/puma/configuration.rb:270 call |  
lib/puma/request.rb:98 block in handle_request |  
lib/puma/thread_pool.rb:340 with_force_shutdown |  
lib/puma/request.rb:97 handle_request |  
lib/puma/server.rb:431 process_client |  
lib/puma/server.rb:233 block in run |  
lib/puma/thread_pool.rb:147 block in spawn_thread


```

</details>

**Additional context**

With these details, we can efficiently hunt down the bug!

**Comment by rmosolgo** [comment]
Hey, sorry for the trouble with this and thanks for the detailed report! I just released graphql-pro v1.24.3 which I expect to fix this problem, could you download that version and give it a try? 

(There was a call to `field.arguments[name]` in `index_visitor.rb` that needed to be updated to `field.get_argument(name, query.context)` instead, so that arguments would be properly filtered by the current changeset.) 

Let me know what you find!

**Comment by mehmetdavut** [comment]
> Hey, sorry for the trouble with this and thanks for the detailed report! I just released graphql-pro v1.24.3 which I expect to fix this problem, could you download that version and give it a try?
> 
> (There was a call to `field.arguments[name]` in `index_visitor.rb` that needed to be updated to `field.get_argument(name, query.context)` instead, so that arguments would be properly filtered by the current changeset.)
> 
> Let me know what you find!

Thank you for help. Now, I'm getting an error with updated gem version:

```json
{
  "added": [],
  "not_modified": [],
  "failed": [
    "f5923476088b2c63166a1bb2c6c7980a"
  ],
  "errors": {
    "f5923476088b2c63166a1bb2c6c7980a": [
      "Type mismatch on variable $required and argument required (OptionalBoolean / Boolean) (2:16)"
    ]
  },
  "committed": false
}
```

The first error (GraphQL::Schema::DuplicateNamesError) was blocking the request and I can't get response. Now, I can get response from operation store but with the error above.

The argument (`required`) was Boolean and we want to modify it to a custom type (OptionalBoolean) by using the changeset.

**Comment by rmosolgo** [comment]
Hmm, alright. Did you include `--changeset-version=...` with your `sync` operation ([doc](https://graphql-ruby.org/javascript_client/sync.html#sync-utility))? That's required to tell the server what version of the schema to use when validating incoming queries. 

If you're already including that, could you add `--verbose` to your sync call and share the output of that command, along with the body of the operation you're syncing? (Feel free to send it by email to rmosolgo@graphql.pro instead of posting it here.)

**Comment by mehmetdavut** [comment]
@rmosolgo sorry for the recent comment, it was my mistake while testing the result. Now, I can confirm that the latest published version of the gem (1.24.3) is successfully fixed the "Found two visible definition" error when modifying an argument in a changeset 👍🏼

**Comment by rmosolgo** [comment]
Glad to hear it, thanks for following up!
